c_cpp_guide.md

```diff
@@ -0,0 +1,68 @@
+
+Quirk 2) In the current ewasm design, all communication between the contract and the client is done through the module's memory. For example, the message data ("call data") to the contract is accessed by calling `callDataCopy` to put this data to a memory location. Accessing a contract's storage requires calls to `storageStore` or `storageLoad`, which require memory offsets of the storage key and value. The memory is allocated in C/C++ using `malloc`. For example, before calling `callDataCopy`, one may use 'getCallDataSize' to see how many bytes of memory to `malloc`.
+
+Quirk 3) In the current ewasm design, the Ethereum client writes data into WebAssembly as big-endian, and WebAssembly memory loads and stores are in little-endian. For example, when the call data is brought into memory using `callDataCopy`, and those bytes are loaded to the WebAssembly stack using `i64.load`, all of the bytes are reversed. So a function may be needed in C/C++ to reverse the bytes.
+
+Quirk 4) The output of compilers is a `.wasm` which may have some extra imports and exports with strange names. These must be manually fixed to be a valid ewasm contract.
+
+
+## Step-by-Step Guide
+
+The wasmception package offers patches to libc and libc++ which allow using `malloc` when targeting WebAssembly. This tool is not specific to ewasm, it is maintained by a third party. Warning: compiling wasmception will download llvm, llvm tools, musl C library, and libc++, which will require internet bandwidth, lots of RAM, and time.
+
+```sh 
+git clone https://github.com/yurydelendik/wasmception.git
+cd wasmception
+make -j4	# Warning: this required lots of bandwidth, RAM, and time. One hours on a mid-level laptop.
+cd ..
+``` 
+
+Write down the end of the output of the above make command, it should be something like: `--sysroot=/home/user/repos/wasmception/sysroot`.
+
+Next we modified the wasmception examples into an ewasm contract, and is available for download. Make sure to edit the Makefile with the sysroot above, and change the path to our newly compiled version of clang to something like `/home/user/repos/wasmception/dist/bin/clang`. Make sure that `main.syms` has a list of ewasm helper functions you are using. If you are using C++, make sure to modify the Makefile to clang++, use `extern "C"` around the helper function declarations, and follow other tips from wasmception.
+
+```sh
+git clone https://gist.github.com/poemm/91b64ecd2ca2f1cb4a88d31315313b9b.git cwrc20
+cd cwrc20
+# edit the Makefile and main.syms as described above
+make
+```
+
+The output is `main.wasm`, but it needs some cleanup to remove some of it's extra import and export junk. For this, we use PyWebAssembly. 
+
+Aside: Alternatively, one can use a rust implementation of `wasm_chisel`. But that requires a rust compiler.
```

Comment:
For python one needs a python interpreter :)

(and it is broken on default mac installation)

Reply:
Hmmm. I want to add a link to wasm chisel in rust. Do you have a link?

Reply:
https://github.com/wasmx/wasm-chisel

Reply:
Installation instruction is though as simple as `cargo install chisel`.